plugin/session_buffer.py

```diff
@@ -344,6 +345,9 @@ def _on_after_change_async(self, view: sublime.View, version: int) -> None:
             return
         self._do_color_boxes_async(view, version)
         self.do_document_diagnostic_async(view, version)
+        if self.session.config.diagnostics_mode == "workspace" and \
+                self.session.has_capability('diagnosticProvider.workspaceDiagnostics'):
+            debounced(self.session.do_workspace_diagnostics_async, WORKSPACE_DIAGNOSTICS_TIMEOUT, async_thread=True)
```

Comment:
Oh, I didn't realize that `debounced` is not really debounced, but instead just "delayed" if there is no condition given. I thought the `debounced` function would already do what `DebouncerNonThreadSafe` does. Perhaps they should be combined into a single one.

And the `_on_after_change_async` is already debounced itself, so I think it would at least not trigger 3 times on typing `abc`, but it's still bad due to the increased timeout for workspace diagnostics.

Reply:
`SessionBuffer.purge_changes_async` is also triggered from `DocumentSyncListener._on_query_completions_async` and due to that, it can in fact be triggered 3 times in that case, depending on whether given context triggers completions. Not sure if this could be optimized.

Reply:
In the case where I saw it it seems it's pyright responding very quickly to all those completion requests (in 40, 20, 15ms) so then it all makes sense.

Reply:
I think it should be better now. I'm not sure also whether the debouncing is really necessary for workspace diagnostics, because they seem to be more of an replacement for the `publishDiagnostics` notification for servers which support pull diagnostics. So it seems that the request might only be sent a single time, and then be kept open forever. Thus I added another condition now, in order to prevent unnecessary debouncing work.